src/flp/types.rs

```diff
@@ -183,27 +160,109 @@ impl<F: FieldElement> Type for Sum<F> {
         _num_shares: usize,
     ) -> Result<F, FlpError> {
         valid_call_check(self, input, joint_rand)?;
+        call_gadget_on_vec_entries(&mut g[0], input, joint_rand[0])
+    }
 
-        // Check that each element of `data` is a 0 or 1.
-        let mut range_check = F::zero();
-        let mut r = joint_rand[0];
-        for chunk in input.chunks(1) {
-            range_check += r * g[0].call(chunk)?;
-            r *= joint_rand[0];
-        }
+    fn truncate(&self, input: Vec<F>) -> Result<Vec<F>, FlpError> {
+        truncate_call_check(self, &input)?;
+        let res = decode_from_bitvector_representation(&input);
+        Ok(vec![res])
+    }
+
+    fn input_len(&self) -> usize {
+        self.bits
+    }
+
+    fn proof_len(&self) -> usize {
+        2 * ((1 + self.bits).next_power_of_two() - 1) + 2
+    }
+
+    fn verifier_len(&self) -> usize {
+        3
+    }
+
+    fn output_len(&self) -> usize {
+        1
+    }
+
+    fn joint_rand_len(&self) -> usize {
+        1
+    }
+
+    fn prove_rand_len(&self) -> usize {
+        1
+    }
+
+    fn query_rand_len(&self) -> usize {
+        1
+    }
+}
+
+/// The average type. Each measurement is an integer in `[0,2^bits)` for some `0 < bits < 64` and the
+/// aggregate is the arithmetic average. This type is a test for demonstrating that the number of measurements
+/// may be used in the `decode_result()` function. No validity checks are done on the input data.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Average<F: FieldElement> {
+    bits: usize,
+    max_summand: F::Integer,
+    range_checker: Vec<F>,
+}
+
+impl<F: FieldElement> Average<F> {
+    /// Return a new [`Average`] type parameter. Each value of this type is an integer in range `[0,
+    /// 2^bits)`.
+    pub fn new(bits: usize) -> Result<Self, FlpError> {
+        let bits_int = valid_int_from_usize::<F>(bits)?;
+        let one = F::Integer::from(F::one());
+        let max_summand = (one << bits_int) - one;
 
-        Ok(range_check)
+        Ok(Self {
+            bits,
+            max_summand,
+            range_checker: poly_range_check(0, 2),
+        })
+    }
+}
+
+impl<F: FieldElement> Type for Average<F> {
+    type Measurement = F::Integer;
+    type AggregateResult = f64;
+    type Field = F;
+
+    fn encode_measurement(&self, summand: &F::Integer) -> Result<Vec<F>, FlpError> {
+        encode_into_bitvector_representation(summand, self.bits, self.max_summand)
+    }
+
+    fn decode_result(&self, data: &[F], num_measurements: usize) -> Result<f64, FlpError> {
+        // Compute the average from the aggregated sum.
+        let data = decode_result(data).unwrap();
+        let data: u64 = data.try_into().unwrap();
+        let result = (data as f64) / (num_measurements as f64);
```

Comment:
Somehow I cannot get your suggestion to work.
 - I think that `data` needs to be a `f64` here, because division seems only to be implemented in the case where both numerator and denominator have the same type. And I assume that we want to use the floating point division here.
 - There is no `from` or `try_from` for `u64 -> f64`, only for `u32 -> f64`, so I think the only other option is to use the "cast with `as`".
 - Similarly there is no `from` for `usize -> f64`.

Could you elaborate on how your comment was meant? Am I missing imports?

Reply:
Ok, this looks fine as is.